src/hotspot/share/code/compressedStream.hpp

```diff
@@ -111,4 +112,93 @@ class CompressedWriteStream : public CompressedStream {
   }
 };
 
+class CompressedBitStream : public ResourceObj {
+protected:
+  u_char* _buffer;
+  int     _position; // current byte offset
+  size_t  _bit_pos {0}; // current bit offset
+
+public:
+  CompressedBitStream(u_char* buffer = NULL, int position = 0) {
+    _buffer   = buffer;
+    _position = position;
+  }
+
+  u_char* buffer() const { return _buffer; }
+};
+
+// Modified compression algorithm for a data set in which a significant part of the data is null
+class CompressedSparseDataReadStream : public CompressedBitStream {
+public:
+  CompressedSparseDataReadStream(u_char* buffer, int position) : CompressedBitStream(buffer, position) {}
+
+  void set_position(int pos) {
+    _bit_pos = 0;
+    _position = pos;
+  }
+
+  jboolean read_bool()       { return read_int(); }
+  jbyte    read_byte()       { return read_int(); }
+  jint     read_signed_int() { return UNSIGNED5::decode_sign(read_int()); }
+  jint     read_int();
+  jdouble  read_double() {
+    jint h = reverse_bits(read_int());
+    jint l = reverse_bits(read_int());
+    return jdouble_cast(jlong_from(h, l));
+  }
+  jlong    read_long() {
+    jint low  = read_signed_int();
+    jint high = read_signed_int();
+    return jlong_from(high, low);
+  }
+
+protected:
+  bool read_zero();
+  uint8_t read_byte_impl();
+  inline u_char read()       { return _buffer[_position++]; }
+};
+
+class CompressedSparseDataWriteStream : public CompressedBitStream {
+public:
+  CompressedSparseDataWriteStream(int initial_size) : CompressedBitStream() {
+    _buffer   = NEW_RESOURCE_ARRAY(u_char, initial_size);
+    _size     = initial_size;
+  }
+
+  void write_bool(jboolean value)   { write_int(value ? 1 : 0); }
+  void write_byte(jbyte value)      { write_int(value); }
+  void write_signed_int(jint value) { write_int(UNSIGNED5::encode_sign(value)); }
+  void write_int(juint value);
+  void write_double(jdouble value)  {
+    juint rh = reverse_bits(high(jlong_cast(value)));
+    juint rl = reverse_bits(low( jlong_cast(value)));
+    write_int(rh);
+    write_int(rl);
+  }
+  void write_long(jlong value)      {
+    write_signed_int(low(value));
+    write_signed_int(high(value));
+  }
+
+  int position(); // method have a side effect: the current byte becomes aligned
+  void set_position(int pos) {
+    position();
+    _position = pos;
```

Comment:
I think `pos` must be `<= position()`. Should we check this?

Reply:
In fact, it is used as rollback. Though I do not want to limit functionality to this usage only.
Let me add `assert(_position < _size, "set_position is only used for rollback");` check here